api/catalog.go

```diff
@@ -185,20 +198,27 @@ func (c *Catalog) Connect(service, tag string, q *QueryOptions) ([]*CatalogServi
 	if tag != "" {
 		tags = []string{tag}
 	}
-	return c.service(service, tags, q, true)
+	return c.service(service, tags, "", q, true)
+}
+
+// ConnectWithFilter is used to query catalog entries for a given Connect-enabled service
+// and returns a subset of the instances which match the given filter expression.
+func (c *Catalog) ConnectWithFilter(service string, filter string, q *QueryOptions) ([]*CatalogService, *QueryMeta, error) {
```

Comment:
Instead of forking all of the api endpoints, would it have made more sense to add the common filter feature into `QueryOptions` instead?

Reply:
It would certainly have been easier although logically it didn't seem like a great fit. All the things currently in there are around caching, staleness or authorization. All the existing filtering was presented as other arguments to functions so I was trying to stick with that.

With that being said, I don't have strong objections to keeping it out of there if everyone else thinks it would be better in there.

Reply:
I'm just thinking of any possible iteration on this sort of thing in the future. What if we found ourselves wanting to provide another common argument to many endpoints, would we have `Connect, ConnectWithFilter, ConnectWithNamespace, ConnectWithFilterAndNamespace`?

If we did come at this with a [functional option](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) approach instead then it runs into conceptual grey areas because we already have one arg called `QueryOptions` already and adding another "options" arg tends to get a bit silly.

If the `filter` were not as pervasive as it is in this PR putting it into `QueryOptions` may be overreach, but it feels like it crossed the pervasiveness threshold.

Reply:
Done